### ISSUE TITLE
make test errors for Cygwin's MinGW64 - unable to load ossltest.dll

### ISSUE THREAD

**Issue by beamer159** [openssl/openssl#7710]
**OS**: Windows 10
**Environment**: Cygwin
**Compiler**: MinGW64

I am getting make test errors when building OpenSSL using MinGW64 provided by Cygwin. I configure with the following:
`perl Configure mingw64 --cross-compile-prefix=x86_64-w64-mingw32-`

Then I make with no problems:
`make`

Then I test:
`make test`

I fail tests such as `test_comp`, `test_key_share`, and `test_sslcertstatus`. The verbose output of `test_sslcertstatus` is at the end of this post. I looked at the output of these tests and they all have one thing in common: they fail to load the shared library `ossltest.dll`. The tests will probably succeed if this library gets loaded correctly. I checked and this file does exist in the location OpenSSL is looking for it.

I would also mention that MinGW provided by MSYS2 works without any problems. I can build OpenSSL with the usual commands:
```
perl Configure mingw64
mingw32-make
mingw32-make test
mingw32-make install
```
---
### test_sslcertstatus output
```
$ make 'TESTS=test_sslcertstatus' VERBOSE=1 test
make depend && make _tests
make[1]: Entering directory '/cygdrive/d/Downloads/openssl-1_1_1a-cygwin'
make[1]: Leaving directory '/cygdrive/d/Downloads/openssl-1_1_1a-cygwin'
make[1]: Entering directory '/cygdrive/d/Downloads/openssl-1_1_1a-cygwin'
( cd test; \
  mkdir -p test-runs; \
  SRCTOP=../. \
  BLDTOP=../. \
  RESULT_D=test-runs \
  PERL="/usr/bin/perl" \
  EXE_EXT=.exe \
  OPENSSL_ENGINES=`cd .././engines 2>/dev/null && pwd` \
  OPENSSL_DEBUG_MEMORY=on \
    /usr/bin/perl .././test/run_tests.pl test_sslcertstatus )
../test/recipes/70-test_sslcertstatus.t ..
Proxy started on port [::1]:62633
Server command: ../../util/shlib_wrap.sh ../../apps/openssl.exe s_server -max_protocol TLSv1.3 -no_comp -rev -engine ossltest -ext_cache -accept [::1]:0 -cert ../../apps/server.pem -cert2 ../../apps/server.pem -naccept 1 -cipher AES128-SHA -ciphersuites TLS_AES_128_GCM_SHA256
Using default temp DH parameters
ACCEPT [::1]:62634
Server responds on [::1]:62634
Client command: ../../util/shlib_wrap.sh ../../apps/openssl.exe s_client -max_protocol TLSv1.3 -engine ossltest -connect [::1]:62633 -status -no_tls1_3 -servername localhost
Connection opened
Received client packet
Packet length = 221
Processing flight 0
 Record 1 (client -> server)
  Content type: HANDSHAKE
  Version: TLS1.0
  Length: 216
  Message type: ClientHello
  Message Length: 212
    Client Version:771
    Session ID Len:0
    Ciphersuite len:56
    Compression Method Len:1
    Extensions Len:115

Forwarded packet length = 221

Received server packet
Packet length = 1093
Processing flight 1
 Record 1 (server -> client)
  Content type: HANDSHAKE
  Version: TLS1.2
  Length: 61
  Message type: ServerHello
  Message Length: 57
    Server Version:771
    Session ID Len:0
    Ciphersuite:47
    Compression Method:0
    Extensions Len:17
 Record 2 (server -> client)
  Content type: HANDSHAKE
  Version: TLS1.2
  Length: 1013
  Message type: Certificate
  Message Length: 1009
    Certificate List Len:1006
    Certificate Len:1003
 Record 3 (server -> client)
  Content type: HANDSHAKE
  Version: TLS1.2
  Length: 4
  Message type: ServerHelloDone
  Message Length: 0

Forwarded packet length = 1097

Received client packet
Packet length = 346
Processing flight 2
 Record 1 (client -> server)
  Content type: HANDSHAKE
  Version: TLS1.2
  Length: 262
  Message type: ClientKeyExchange
  Message Length: 258
 Record 2 (client -> server)
  Content type: CCS
  Version: TLS1.2
  Length: 1
 Record 3 (client -> server)
  Content type: HANDSHAKE
  Version: TLS1.2
  Length: 68

Forwarded packet length = 346

Received server packet
Packet length = 7
Processing flight 3
 Record 1 (server -> client)
  Content type: ALERT
  Version: TLS1.2
  Length: 2
  [2, 20]

Forwarded packet length = 7

CONNECTED(0000010C)
OCSP response: no response sent
---
Certificate chain
 0 s:C = UK, O = OpenSSL Group, OU = FOR TESTING PURPOSES ONLY, CN = Test Server Cert
   i:C = UK, O = OpenSSL Group, OU = FOR TESTING PURPOSES ONLY, CN = OpenSSL Test Intermediate CA
---
Server certificate
-----BEGIN CERTIFICATE-----
MIID5zCCAs+gAwIBAgIJALnu1NlVpZ6zMA0GCSqGSIb3DQEBBQUAMHAxCzAJBgNV
BAYTAlVLMRYwFAYDVQQKDA1PcGVuU1NMIEdyb3VwMSIwIAYDVQQLDBlGT1IgVEVT
VElORyBQVVJQT1NFUyBPTkxZMSUwIwYDVQQDDBxPcGVuU1NMIFRlc3QgSW50ZXJt
ZWRpYXRlIENBMB4XDTExMTIwODE0MDE0OFoXDTIxMTAxNjE0MDE0OFowZDELMAkG
A1UEBhMCVUsxFjAUBgNVBAoMDU9wZW5TU0wgR3JvdXAxIjAgBgNVBAsMGUZPUiBU
RVNUSU5HIFBVUlBPU0VTIE9OTFkxGTAXBgNVBAMMEFRlc3QgU2VydmVyIENlcnQw
ggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQDzhPOSNtyyRspmeuUpxfNJ
KCLTuf7g3uQ4zu4iHOmRO5TQci+HhVlLZrHF9XqFXcIP0y4pWDbMSGuiorUmzmfi
R7bfSdI/+qIQt8KXRH6HNG1t8ou0VSvWId5TS5Dq/er5ODUr9OaaDva7EquHIcMv
vPQGuI+OEAcnleVCy9HVEIySrO4P3CNIicnGkwwiAud05yUAq/gPXBC1hTtmlPD7
TVcGVSEiJdvzqqlgv02qedGrkki6GY4S7GjZxrrf7Foc2EP+51LJzwLQx3/JfrCU
41NEWAsu/Sl0tQabXESN+zJ1pDqoZ3uHMgpQjeGiE0olr+YcsSW/tJmiU9OiAr8R
AgMBAAGjgY8wgYwwDAYDVR0TAQH/BAIwADAOBgNVHQ8BAf8EBAMCBeAwLAYJYIZI
AYb4QgENBB8WHU9wZW5TU0wgR2VuZXJhdGVkIENlcnRpZmljYXRlMB0GA1UdDgQW
BBSCvM8AABPR9zklmifnr9LvIBturDAfBgNVHSMEGDAWgBQ2w2yI55X+sL3szj49
hqshgYfa2jANBgkqhkiG9w0BAQUFAAOCAQEAqb1NV0B0/pbpK9Z4/bNjzPQLTRLK
WnSNm/Jh5v0GEUOE/Beg7GNjNrmeNmqxAlpqWz9qoeoFZax+QBpIZYjROU3TS3fp
yLsrnlr0CDQ5R7kCCDGa8dkXxemmpZZLbUCpW2Uoy8sAA4JjN9OtsZY7dvUXFgJ7
vVNTRnI01ghknbtD+2SxSQd3CWF6QhcRMAzZJ1z1cbbwGDDzfvGFPzJ+Sq+zEPds
xoVLLSetCiBc+40ZcDS5dV98h9XD7JMTQfxzA7mNGv73JoZJA6nFgj+ADSlJsY/t
JBv+z1iQRueoh9Qeee+ZbRifPouCB8FDx+AltvHTANdAq0t/K3o+pplMVA==
-----END CERTIFICATE-----
subject=C = UK, O = OpenSSL Group, OU = FOR TESTING PURPOSES ONLY, CN = Test Server Cert

issuer=C = UK, O = OpenSSL Group, OU = FOR TESTING PURPOSES ONLY, CN = OpenSSL Test Intermediate CA

---
No client certificate CA names sent
---
SSL handshake has read 1104 bytes and written 567 bytes
Verification error: unable to verify the first certificate
---
New, SSLv3, Cipher is AES128-SHA
Server public key is 2048 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : AES128-SHA
    Session-ID:
    Session-ID-ctx:
    Master-Key: CED0BF88BDF14B5BE4FFA1228EF259D5CB9AD9901BDD70C726D29F729A4A289D31B559EB12D28DE0ADD768C7874D7E06
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    Start Time: 1543299837
    Timeout   : 7200 (sec)
    Verify return code: 21 (unable to verify the first certificate)
    Extended master secret: yes
---
Connection closed
Waiting for 'perl -ne print' process to close: 7208...
invalid engine "ossltest"
11184:error:25078067:DSO support routines:win32_load:could not load the shared library:crypto/dso/dso_win32.c:107:filename(\cygdrive\d\Downloads\openssl-1_1_1a-cygwin\engines\ossltest.dll)
11184:error:25070067:DSO support routines:DSO_load:could not load the shared library:crypto/dso/dso_lib.c:162:
11184:error:260B6084:engine routines:dynamic_load:dso not found:crypto/engine/eng_dyn.c:414:
11184:error:2606A074:engine routines:ENGINE_by_id:no such engine:crypto/engine/eng_list.c:334:id=ossltest
11184:error:25078067:DSO support routines:win32_load:could not load the shared library:crypto/dso/dso_win32.c:107:filename(ossltest.dll)
11184:error:25070067:DSO support routines:DSO_load:could not load the shared library:crypto/dso/dso_lib.c:162:
11184:error:260B6084:engine routines:dynamic_load:dso not found:crypto/engine/eng_dyn.c:414:
depth=0 C = UK, O = OpenSSL Group, OU = FOR TESTING PURPOSES ONLY, CN = Test Server Cert
verify error:num=20:unable to get local issuer certificate
verify return:1
depth=0 C = UK, O = OpenSSL Group, OU = FOR TESTING PURPOSES ONLY, CN = Test Server Cert
verify error:num=21:unable to verify the first certificate
verify return:1
11184:error:140943FC:SSL routines:ssl3_read_bytes:sslv3 alert bad record mac:ssl/record/rec_layer_s3.c:1528:SSL alert number 20
   0 items in the session cache
   0 client connects (SSL_connect())
   0 client renegotiates (SSL_connect())
   0 client connects that finished
   1 server accepts (SSL_accept())
   0 server renegotiates (SSL_accept())
   0 server accepts that finished
   0 session cache hits
   0 session cache misses
   0 session cache timeouts
   0 callback cache hits
   0 cache full overflows (20480 allowed)
invalid engine "ossltest"
13832:error:25078067:DSO support routines:win32_load:could not load the shared library:crypto/dso/dso_win32.c:107:filename(\cygdrive\d\Downloads\openssl-1_1_1a-cygwin\engines\ossltest.dll)
13832:error:25070067:DSO support routines:DSO_load:could not load the shared library:crypto/dso/dso_lib.c:162:
13832:error:260B6084:engine routines:dynamic_load:dso not found:crypto/engine/eng_dyn.c:414:
13832:error:2606A074:engine routines:ENGINE_by_id:no such engine:crypto/engine/eng_list.c:334:id=ossltest
13832:error:25078067:DSO support routines:win32_load:could not load the shared library:crypto/dso/dso_win32.c:107:filename(ossltest.dll)
13832:error:25070067:DSO support routines:DSO_load:could not load the shared library:crypto/dso/dso_lib.c:162:
13832:error:260B6084:engine routines:dynamic_load:dso not found:crypto/engine/eng_dyn.c:414:
CONNECTION FAILURE
13832:error:1408F119:SSL routines:ssl3_get_record:decryption failed or bad record mac:ssl/record/ssl3_record.c:521:
Waiting for s_server process to close: 1520...
Waiting for s_client process to close: 12732...
1..1
not ok 1 - Missing CertificateStatus message

#   Failed test 'Missing CertificateStatus message'
#   at ../test/recipes/70-test_sslcertstatus.t line 46.
# Looks like you failed 1 test of 1.
Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests

Test Summary Report
-------------------
../test/recipes/70-test_sslcertstatus.t (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
Files=1, Tests=1,  0 wallclock secs ( 0.02 usr  0.05 sys +  0.20 cusr  0.42 csys =  0.68 CPU)
Result: FAIL
make[1]: *** [Makefile:205: _tests] Error 1
make[1]: Leaving directory '/cygdrive/d/Downloads/openssl-1_1_1a-cygwin'
make: *** [Makefile:202: tests] Error 2
```

**Comment by levitte** [comment]
Without currently having access to my Windows equipment, I'm *guessing* that the there's a clash of expectations here.  Mingw executables are really just Windows executables, so they usually expect Windows style paths, which `\cygdrive\d\Downloads\openssl-1_1_1a-cygwin\engines\ossltest.dll` isn't (they would rather expect `D:\Downloads\openssl-1_1_1a-cygwin\engines\ossltest.dll`).  Since Cygwin works differently in this regard, the path doesn't get converted in a way that a mingw executable would normally expect (there's a bit of magic around that that I have still not quite wrapped my head around).

Bottom line is that you're running a combination of things that we have very little support for, and it's no surprise that the MSYS2/Mingw combination works for you, since that's a combination we do support (and regularly test).
